bitcoin/bitcoin
# kernel, refactor: Replace `goto` with RAII in `bitcoin-chainstate`

This PR replaces the [`goto` statement](https://github.com/bitcoin/bitcoin/pull/24304/files#diff-04e685224f1ac5bfd91d47d8d7528a2e44f94fab5535d4b6b5af79b5a13aeb93R91) in `bitcoin-chainstate` by encapsulating resources into a dedicated `Epilogue` class.
I've bumped into it while reviewing [#25722](https://github.com/bitcoin/bitcoin/pull/25722/files#diff-04e685224f1ac5bfd91d47d8d7528a2e44f94fab5535d4b6b5af79b5a13aeb93L132).

Key Changes:
* Encapsulated `ValidationSignals` and `ChainstateManager` cleanup into an RAII Epilogue class to replace the goto-based cleanup logic.
* Split `bitcoin-chainstate` logic into a header to enable isolated testing and clarity.
* Introduced basic tests to ensure normal execution completes successfully in regtest, and cleanup code is called in both success and failure scenarios.

---

How to test it manually:
* `cmake -B build -DCMAKE_BUILD_TYPE=Debug -DBUILD_UTIL_CHAINSTATE=OFF && cmake --build build -j$(nproc) && ctest --test-dir build -j$(nproc)` - should not contain `bitcoin_chainstate_tests`
* `cmake -B build -DCMAKE_BUILD_TYPE=Debug -DBUILD_UTIL_CHAINSTATE=ON && cmake --build build -j$(nproc) && ctest --test-dir build -j$(nproc)` - should contain a passing `bitcoin_chainstate_tests`
* `cmake -B build -DCMAKE_BUILD_TYPE=Debug -DBUILD_UTIL_CHAINSTATE=ON && cmake --build build -j$(nproc) && ./build/src/bitcoin-chainstate your-bitcoin-datadir` - should work the same as before (needs an extra newline at the end to exit)
